Disable captcha storefont config still load captcha js files

rogerdz opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#32987

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. ...
2. ...

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

rogerdz committed:
Disable captcha storefont config still load captcha js files

rogerdz commented:
@magento run all tests

rogerdz committed:
update code

rogerdz committed:
Merge branch '2.4-develop' into fix-32987

rogerdz commented:
@magento run all tests

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-32987

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented on app/code/Magento/Captcha/Plugin/CheckCaptchaOnStorefront.php:
```diff
@@ -0,0 +1,75 @@
+     *
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function afterGetTemplate(
```
Please specify return type declaration.

engcom-Hotel commented on app/code/Magento/Captcha/Plugin/CheckCaptchaOnStorefront.php:
```diff
@@ -0,0 +1,75 @@
+     */
+    public function afterGetTemplate(
+        AuthenticationPopup $subject,
+        $result
```
Please add parameter type declaration.

engcom-Hotel commented on app/code/Magento/Captcha/Plugin/CheckCaptchaOnStorefront.php:
```diff
@@ -0,0 +1,75 @@
+     *
+     * @return bool
+     */
+    private function isLoggedIn()
```
Please specify return type declaration.

engcom-Hotel commented on app/code/Magento/Captcha/Plugin/CheckCaptchaOnStorefront.php:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    /**
+     * Remove template when loggin or disable captcha storefront
```
Correct the typo from `loggin` to `login`

engcom-Hotel commented on app/code/Magento/Captcha/Plugin/CheckCaptchaOnStorefront.php:
```diff
@@ -0,0 +1,75 @@
+     * @param HttpContext $httpContext
+     */
+    public function __construct(
+        HelperCaptcha $helper,
```
Use [readonly](https://php.watch/versions/8.1/readonly#constructor-readonly) property here

engcom-Hotel reviewed (changes requested):
Hello @rogerdz,

Thank you for the contribution!

Please address the review comments mentioned below.Additionally it appears that the failed integration and static tests are caused by the changes in this PR. 

Kindly resolve those as well.

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-32987

rogerdz commented:
> Hello @rogerdz,
> 
> Thank you for the contribution!
> 
> Please address the review comments mentioned below.Additionally it appears that the failed integration and static tests are caused by the changes in this PR.
> 
> Kindly resolve those as well.

not yet, please fix your composer repo first
![image](https://github.com/user-attachments/assets/45148a41-e30a-4ce5-9f68-3105365d008f)

engcom-Hotel commented:
Hello @rogerdz,

This is a known issue and there is an ongoing discussion on this [here](https://magentocommeng.slack.com/archives/C7KB93M32/p1736933306053669). Please refer to this chat for further information.

In the meantime, do you want us to continue with this PR, or should we wait for the composer-related issue to be fixed and then you continue to work on this issue?

Thanks

rogerdz commented:
> do you want us to continue with this PR

Yes, thank you.

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-32987

engcom-Hotel committed:
Fix the static test failures and improve code readability

engcom-Hotel commented:
@magento run all tests

engcom-Hotel committed:
Fix static test failures

engcom-Hotel commented:
@magento run Static Tests

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-32987

engcom-Hotel commented:
@magento run all tests

glo24157 committed:
Fix integration test failures

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-32987

engcom-Hotel commented:
@magento run all tests

glo24157 committed:
Fixes integration test failures

engcom-Hotel committed:
Fixes static test failures

engcom-Hotel commented:
@magento run Integration Tests, Static Tests

engcom-Hotel commented:
@magento run all tests

ihor-sviziev commented on app/code/Magento/Captcha/Plugin/CheckCaptchaOnStorefront.php:
```diff
@@ -0,0 +1,71 @@
+     *
+     * @return bool
+     */
+    private function isLoggedIn(): ?bool
+    {
+        return $this->httpContext->getValue(Context::CONTEXT_AUTH);
+    }
```
I guess this check should be in the original class `\Magento\Customer\Block\Account\AuthenticationPopup`, or in the template file `Magento_Customer::account/authentication-popup.phtml`.

ihor-sviziev commented on app/code/Magento/Captcha/Plugin/CheckCaptchaOnStorefront.php:
```diff
@@ -0,0 +1,71 @@
+            ScopeInterface::SCOPE_WEBSITE
+        );
+
+        if ($this->isLoggedIn() || (!$this->helper->getConfig('enable') && $recaptchaType === null)) {
```
I guess the logic about enabled form for custom login should be added at place, where we configure captcha fields in the customer login form.

I checked - currently it's configured using layout XML, so I guess it has to be moved from XML to some PHP code that will add needed parameters. https://github.com/magento/magento2/blob/c5614048eaaff80c1d4786b391e0c3cc55c7417c/app/code/Magento/Captcha/view/frontend/layout/checkout_index_index.xml#L12-L56

----

I guess it should be something like this:

https://github.com/magento/magento2/blob/c5614048eaaff80c1d4786b391e0c3cc55c7417c/app/code/Magento/Checkout/Block/Onepage.php#L78-L85
https://github.com/magento/magento2/blob/c5614048eaaff80c1d4786b391e0c3cc55c7417c/app/code/Magento/Paypal/etc/frontend/di.xml#L197-L203
https://github.com/magento/magento2/blob/c5614048eaaff80c1d4786b391e0c3cc55c7417c/app/code/Magento/Paypal/Block/PayLater/LayoutProcessor.php#L20-L68

ihor-sviziev reviewed (changes requested):
Hi @rogerdz @engcom-Hotel,
I believe the captcha module should not affect customer functionality. It is better to adjust the original functionality rather than introduce new plugins.

engcom-Hotel committed:
Update code according to the review comment

Handle Captcha JS using LayoutProcessor for checkout

Add a condition in customer login to prevent captcha related JS

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-32987

engcom-Hotel commented:
@magento run all tests

engcom-Hotel committed:
Fixed static test failures

engcom-Hotel commented:
@magento run all tests

hostep commented on app/code/Magento/Customer/Block/Account/AuthenticationPopup.php:
```diff
@@ -24,21 +27,29 @@ class AuthenticationPopup extends \Magento\Framework\View\Element\Template
         array $data = [],
-        ?\Magento\Framework\Serialize\Serializer\Json $serializer = null
+        ?\Magento\Framework\Serialize\Serializer\Json $serializer = null,
+        HttpContext $httpContext = null
```
Can we update this type so it's explicitly nullable, this is a requirement for PHP 8.4 to not get deprecation errors.
So change `HttpContext` to `?HttpContext` please. Thanks! :) 

----

Wierd that static tests don't see this, or don't we check on PHP 8.4 compatibility yet?

_Update_, found reasons why:
- phpstan only complains about this when we execute it with PHP 8.4 and not with PHP 8.3, which kind of makes sense, so should we start running phpstan with all the versions of PHP supported in the codebase in the future?
- the PHPCompatibility ruleset from magento-coding-standards is very outdated, I've opened a new issue here about it: https://github.com/magento/magento-coding-standard/issues/491

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-32987

engcom-Hotel committed:
Fix review comment

handling nullable type

engcom-Hotel commented:
@magento run all tests

ihor-sviziev commented on app/code/Magento/Captcha/Block/CaptchaLayoutProcessor.php:
```diff
@@ -0,0 +1,48 @@
+     * @param array $jsLayout
+     * @return array
+     */
+    public function process($jsLayout): array
```
```suggestion
    public function process(array $jsLayout): array
```

engcom-Hotel replied:
Add `array` here showing error in PHPStorm because the interface doesn’t have this declaration. The error shown in PHPStorm is as follows:
![image](https://github.com/user-attachments/assets/82e47aa9-90d0-457d-b3ee-78d26aa11e98)

ihor-sviziev commented on app/code/Magento/Customer/Block/Account/AuthenticationPopup.php:
```diff
@@ -48,6 +59,16 @@ public function __construct(
+            if (isset($this->jsLayout['components']['authenticationPopup']['children']['captcha'])) {
+                unset($this->jsLayout['components']['authenticationPopup']['children']['captcha']);
+            }
+        }
```
The logic you've just added is related to the captha extension, as it's added in that module in https://github.com/magento/magento2/blob/df37b871ebd2a7e23ae6dec8c084855a65fbcab1/app/code/Magento/Captcha/view/frontend/layout/default.xml#L10-L27

I would suggest doing the following:
1. add layoutProcessors as optional array option to constructor
2. Adjust getJsLayout function to something like this
```suggestion
        foreach ($this->layoutProcessors as $processor) { 
            $this->jsLayout = $processor->process($this->jsLayout); 
        }
```
This will make code as close as possible to https://github.com/magento/magento2/blob/c5614048eaaff80c1d4786b391e0c3cc55c7417c/app/code/Magento/Checkout/Block/Onepage.php#L78-L85
3. In the Capcha extension, create own layout processor for AuthenticationPopup and **add** all needed parameters through this class + remove the configuration code from layout.xml

**OR**
Do some plugin for this specific method in the captcha extension, and add needed parameters based on the configuration

ihor-sviziev commented on app/code/Magento/Captcha/Block/CaptchaLayoutProcessor.php:
```diff
@@ -0,0 +1,48 @@
+            if (isset($jsLayout['components']['authenticationPopup']['children']['captcha'])) {
+                unset($jsLayout['components']['authenticationPopup']['children']['captcha']);
+            }
+        }
```
I would suggest adding needed parameters rather than removing them (move from layout.xml), similar to 
https://github.com/magento/magento2/blob/b25fc6f39f584d527745b9d0149c5018c6854d9d/app/code/Magento/Paypal/Block/PayLater/LayoutProcessor.php#L70-L105

ihor-sviziev commented on app/code/Magento/Customer/Block/Account/AuthenticationPopup.php:
```diff
@@ -48,6 +59,16 @@ public function __construct(
+        if (!$this->_scopeConfig->getValue(
+            Form::XML_PATH_CUSTOMER_CAPTCHA_ENABLED,
+            ScopeInterface::SCOPE_STORE
+        ) && !$this->isLoggedIn()) {
```
I don't think it can be related to whether the customer logged in or not. I would expect that this entire block won't be rendered then, but that's a separate issue.

ihor-sviziev commented on app/code/Magento/Captcha/etc/frontend/di.xml:
```diff
@@ -37,4 +37,12 @@
+    <type name="Magento\Checkout\Block\Onepage">
+        <arguments>
+            <argument name="layoutProcessors" xsi:type="array">
+                <item name="captcha_processor" xsi:type="object">Magento\Captcha\Block\CaptchaLayoutProcessor</item>
```
I guess from the capthca module perspective, it's more "CheckoutLayoutProcessor" than "CaptchaLayoutProcessor".

engcom-Hotel committed:
Fix review comments

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-32987

engcom-Hotel committed:
Add Layout Processor

engcom-Hotel committed:
Removed some file and unused code

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented:
@magento run Functional Tests EE, Functional Tests B2B

ihor-sviziev commented on app/code/Magento/Captcha/view/frontend/layout/default.xml:
```diff
@@ -13,12 +13,7 @@
-                                    <item name="formId" xsi:type="string">user_login</item>
-                                    <item name="configSource" xsi:type="string">checkout</item>
-                                </item>
+
```
I guess we can remove this file now, and the result will be the same as the child element configured in the PHP file now, right?

ihor-sviziev commented on app/code/Magento/Captcha/view/frontend/layout/checkout_index_index.xml:
```diff
@@ -15,12 +15,7 @@
-                                            <item name="formId" xsi:type="string">user_login</item>
-                                            <item name="configSource" xsi:type="string">checkoutConfig</item>
-                                        </item>
+
```
I guess we can remove this file now, and the result will be the same as the child element configured in the PHP file now, right?

ihor-sviziev commented on app/code/Magento/Captcha/view/frontend/layout/checkout_index_index.xml:
```diff
@@ -33,12 +28,7 @@
-                                                                            <item name="formId" xsi:type="string">user_login</item>
-                                                                            <item name="configSource" xsi:type="string">checkoutConfig</item>
-                                                                        </item>
+
```
same here, it's not adding any value

ihor-sviziev commented on app/code/Magento/Captcha/etc/frontend/di.xml:
```diff
@@ -37,4 +37,18 @@
+    <type name="Magento\Checkout\Block\Onepage">
+        <arguments>
+            <argument name="layoutProcessors" xsi:type="array">
+                <item name="checkout_processor" xsi:type="object">Magento\Captcha\Block\CheckoutLayoutProcessor</item>
```
In terms of checkout onepage block, it's more related to captaha, so let's change the name here
```suggestion
                <item name="captcha_processor" xsi:type="object">Magento\Captcha\Block\CheckoutLayoutProcessor</item>
```

ihor-sviziev commented on app/code/Magento/Captcha/etc/frontend/di.xml:
```diff
@@ -37,4 +37,18 @@
+    <type name="Magento\Customer\Block\Account\AuthenticationPopup">
+        <arguments>
+            <argument name="layoutProcessors" xsi:type="array">
+                <item name="captcha_processor" xsi:type="object">Magento\Captcha\Block\Checkout\LayoutProcessor</item>
```
The current name of the class is quite confusing:
- it's almost the same as  `Magento\Captcha\Block\CheckoutLayoutProcessor`, the differnce is the only one bachslash
- It's not linked anyhow to customer and to authentication popup
So I'm suggesting renaming the class/namespace + update it in di.xml to have something like this
```suggestion
                <item name="captcha_processor" xsi:type="object">Magento\Captcha\Block\Customer\AuthenticationPopup\LayoutProcessor</item>
```

ihor-sviziev commented on app/code/Magento/Customer/Block/Account/AuthenticationPopup.php:
```diff
@@ -24,21 +27,37 @@ class AuthenticationPopup extends \Magento\Framework\View\Element\Template
         array $data = [],
-        ?\Magento\Framework\Serialize\Serializer\Json $serializer = null
+        ?\Magento\Framework\Serialize\Serializer\Json $serializer = null,
+        ?HttpContext $httpContext = null,
```
this parameter isn't used anymore, pls remove

ihor-sviziev commented on app/code/Magento/Customer/Model/Form.php:
```diff
@@ -15,7 +15,7 @@ class Form extends \Magento\Eav\Model\Form
      */
     public const XML_PATH_ENABLE_AUTOCOMPLETE = 'customer/password/autocomplete_on_storefront';
-
+    
```
Pls revert this change
```suggestion

```

engcom-Hotel committed:
Fixes review comments

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-32987

engcom-Hotel commented:
@magento run all tests

ihor-sviziev commented on app/code/Magento/Customer/Block/Account/AuthenticationPopup.php:
```diff
@@ -24,21 +27,34 @@ class AuthenticationPopup extends \Magento\Framework\View\Element\Template
      */
     private $serializer;
 
+    /**
+     * @var HttpContext
+     */
+    private $httpContext;
+
```
```suggestion
```

engcom-Hotel committed:
Fix static test failures

engcom-Hotel commented:
Hello @ihor-sviziev,

Thanks for approving the PR. Moving it to `ready for testing` bucket.

@magento run all tests

engcom-Dash committed:
Merge branch '2.4-develop' into fix-32987

engcom-Dash commented:
@magento run all tests

engcom-Dash commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Unit Tests
